### PR TITLE
Unicode support Part 1

### PR DESCRIPTION
**Pull request by MrPerson**
## About The Pull Request

Makes all calls to ascii2text() and text2ascii() unicode aware as well as all code in code/__HELPERS/text.dm.

Adds defines for the footext_char() procs to maintain 512 support.

I did some cleanup on the text helpers since a lot of them were really, really bad.

reject_bad_text() and reject_bad_name() have an arg to reject non-ascii chars which defaults to TRUE.

## Why It's Good For The Game

Gonna need support for those eggplant emojis... and I guess Cyrillic

### CODE REVIEW

**Comment by RaveRadbury** [comment]
Is this going to fix things like "John Doe&#39;s Awesome Paystand"?

**Comment by MrPerson** [comment]
Probably not, that kind of thing is down to html encoding and decoding, not reading actual unicode.

**Comment by MrPerson** [comment]
I don't know what to do about the failed check, seems above my pay grade.

**Comment by TheVekter** [comment]
> I don't know what to do about the failed check, seems above my pay grade.

You're getting undefined proc errors when it's compiled.

**Comment by SpaceManiac** [comment]
argh, now I'm going to have to rebuke all these uses of the `_char` functions

File: code/__HELPERS/text.dm
Comment (AnturK) [inline code/__HELPERS/text.dm]: If you know it's space you don't really need _char right ?

**Review by AnturK — COMMENTED** [review]
Time to get on encoding pain train.

File: code/__HELPERS/text.dm
Comment (MrPerson) [reply code/__HELPERS/text.dm]: http://jkorpela.fi/chars/spaces.html

There are spaces besides " " char 32.

Comment (AnturK) [reply code/__HELPERS/text.dm]: But you don't check for these.

Comment (MrPerson) [reply code/__HELPERS/text.dm]: Well yeah but if someone ever does want to, now they don't have to change this line or, more likely, forget to.

**Comment by SpaceManiac** [comment]
This needs to be thoroughly reviewed, I'm not at all sold on the `_char` proc family

**Comment by MrPerson** [comment]
Sold on it working correctly or sold on it being needed at all? Like I'm gonna be up front here: I literally didn't know anything about UTF-8 a week ago, so you're probably right.

File: code/__HELPERS/text.dm
Comment (SpaceManiac) [inline code/__HELPERS/text.dm]: Correct. :+1:

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
I did not finish reviewing `code/__HELPERS/text.dm` but this should give you some idea of why I disapprove of the `_char` functions

**Comment by MrPerson** [comment]
I'm getting mixed signals on if we should favor the byte-based or char-based procs. I'm trying to split the difference and use byte-based counts when looping over the entire string or when it was easy and char-based ones when necessary.

Listen we can clean this up later, I'm just trying to get things functional, that's all.

File: code/game/objects/items/devices/PDA/cart.dm
Comment (MrPerson) [reply code/game/objects/items/devices/PDA/cart.dm]: Outside the scope for this tbh. If you want I can just get rid of the copytext outright but either way I'm not getting into this kind of stuff right now.

File: code/modules/reagents/chemistry/reagents/other_reagents.dm
Comment (MrPerson) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm]: I would still have to call i += length(string[i]) so there's no improvement there

**Comment by SpaceManiac** [comment]
> I'm getting mixed signals

If it's another maintainer here let me know and I'll hash it out with them.

I respect the effort but I'd rather we do this right than rely on cheap stopgaps.

File: code/game/objects/items/devices/PDA/cart.dm
Comment (SpaceManiac) [reply code/game/objects/items/devices/PDA/cart.dm]: Fine to punt this out of scope, but I wanted to point out that `_char` does not help

**Comment by MrPerson** [comment]
At this point everyone would just defer to you anyway since you clearly know what you're talking about.

If by "cheap stopgags" you mean "just replace all text procs with _char versions everywhere", I agree that's dumb and I would love to not call _char procs at all. I'll try to think of ways to get rid of more of them.

File: code/_compile_options.dm
Comment (ninjanomnom) [reply code/_compile_options.dm]: Rapid compatibility breakages is exactly what beta releases are for

File: code/game/objects/items/devices/PDA/cart.dm
Comment (MrPerson) [reply code/game/objects/items/devices/PDA/cart.dm]: You know what, I'll toss in a comment saying as such

**Comment by MrPerson** [comment]
That twitterize() proc's use of copytext_char() could probably be rewritten using regex but what's here is functional. The rest of the length_char() calls are kind of wrong for the same reason the PDA-area name use case is wrong, but there's not much to do about that now.

**Comment by optimumtact** [comment]
the reality is that most of the text handling SHOULD be replaced by regex use, rather than trying to treat strings like arrays of characters

Where that is not possible we should use the _char forms for pragmatism

File: code/_compile_options.dm
Comment (Cyberboss) [inline code/_compile_options.dm]: Typo

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Pull request merged**
